File: 0056-merge-intervals/0056-merge-intervals.cpp
class Solution {
public:
    static bool HelperSort(vector<int>& a, vector<int>& b) {
        return a[0] < b[0]; 
    } 

    vector<vector<int>> merge(vector<vector<int>>& nums) {
        int n = nums.size();
        vector<vector<int>> ans;
        sort(nums.begin(), nums.end(), HelperSort);
        int i = 0;
        // while(i<n) {
        //     if(i+1 < n && nums[i][1] >= nums[i+1][0]) {
        //         res.push_back({nums[i][0], max(nums[i][1], nums[i+1][1])});i+=1;
        //     }
        //     else {
        //         res.push_back({nums[i][0], nums[i][1]});i+=1;
        //     }
        // }

        for(int i=0; i<n; i++) {
            if(ans.empty() || ans.back()[1] < nums[i][0]) {
                ans.push_back(nums[i]);
            }
            else {
                ans.back()[1] = max(ans.back()[1], nums[i][1]);
            }
        }
        return ans;
    }
};

File: 0056-merge-intervals/0056-merge-intervals.py
class Solution:
    def merge(self, arr: List[List[int]]) -> List[List[int]]:
        res = []
        # for i in range(1, n):
        #     if not res or (res[i-1][1] >= arr[i][0]):
        #         res.append([res[i-1][0], arr[i][1]])
        #     else:
        #         res.append(arr[i])

        arr.sort(key = lambda x: x[0])

        for interval in arr:
            if not res or res[-1][1] < interval[0]:
                res.append(interval)
            else:
                res[-1][1] = max(res[-1][1], interval[1])
            
        return res